CDPCP-1941. Sync Azure OID Mappings as part of user sync

aarman-cloudera opened the pull request:
Azure cloud identity information (Azure OID mapping) are now synced from
UMS into all appropriate datalake instances. At high level the
implementation is:

- SDX service has a new endpoint that allows setting azure ranger
configs (azure user and group mappings). The SDX service will make the
necessary CM api call (through cluster proxy) to set the config and
trigger a role refresh.

- The FMS user sync service will pull the Azure Object IDs (cloud
identity mapping) from UMS and call the new SDX endpoint to set the
configs.

Follow on work items include:
- Poll the status of the refresh after triggering it
- Individual user level and group level sync of OIDs
- Moving some of the CM logic into cluster-cm and client-cm

handavid commented on datalake/src/main/java/com/sequenceiq/datalake/cm/CloudIdentityRangerSyncService.java:
```diff
@@ -0,0 +1,41 @@
+import java.util.Map;
+
+@Service
+public class CloudIdentityRangerSyncService {
```
nit: I think a better name would be RangerCloudIdentityService

handavid commented on datalake-api/src/main/java/com/sequenceiq/sdx/api/endpoint/SdxEndpoint.java:
```diff
@@ -165,4 +166,10 @@ SdxClusterResponse create(@ValidStackNameFormat @ValidStackNameLength @PathParam
     List<AdvertisedRuntime> advertisedRuntimes();
 
+    @POST
+    @Path("/crn/{crn}/ranger_cloud_identity_mapping")
```
it looks like you want to do this by env crn instead of cluster crn. You should change the path and the parameter name. see getByEnvCrn method above for comparison.

handavid commented on datalake/src/main/java/com/sequenceiq/datalake/cm/ClouderaManagerRangerUtil.java:
```diff
@@ -0,0 +1,124 @@
+        return config;
+    }
+
+    @Retryable(value = RetryException.class, maxAttempts = 5, backoff = @Backoff(delay = 500))
```
why is this the only API that needs retrying?

aarman-cloudera replied:
Refresh specifically might need retrying because another refresh is occurring. Though I've found that there is automatic retry of failed api calls, so I don' think a retry here is necessary anymore. I've removed it.

handavid commented on freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/UmsUsersStateProvider.java:
```diff
@@ -217,11 +219,26 @@ private void addMemberToInternalTrackingGroup(UsersState.Builder usersStateBuild
         fmsUser.withName(umsUser.getWorkloadUsername());
         fmsUser.withFirstName(getOrDefault(umsUser.getFirstName(), "None"));
         fmsUser.withLastName(getOrDefault(umsUser.getLastName(), "None"));
+        fmsUser.withAzureObjectId(getOptionalAzureObjectId(umsUser.getCloudIdentitiesList()));
```
The AzureObjectId belongs in the UmsUsersState, not the UsersState because it's information that is provided by the UMS with no matching information from the FreeIPA side. We use equality to compare the FmsUser and FmsGroup models generated by the UMS against those generated by FreeIPA. Adding in asymmetric information will break the equality and the state difference calculation.
see how we handle workload credentials instead. I think that you can pass the List<CloudIdentity> into the handleUser method and store it there. This will also let you only store OIDs for users that have access to the environment instead of all users.

aarman-cloudera replied:
Thanks, I've moved cloud identify information into the UmsUserState.

handavid commented on freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/UmsUsersStateProvider.java:
```diff
@@ -241,12 +258,14 @@ private FmsUser umsMachineUserToUser(MachineUser umsMachineUser) {
     private FmsGroup umsGroupToGroup(Group umsGroup) {
         FmsGroup fmsGroup = new FmsGroup();
         fmsGroup.withName(umsGroup.getGroupName());
+        fmsGroup.withAzureObjectId(getOptionalAzureObjectId(umsGroup.getCloudIdentitiesList()));
```
FmsGroups suffer the same issue as FmsUsers.
you'll need to calculate the group mapping somewhere around line 71 and add it into the UmsUsersState somewhere around line 78

handavid commented on freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/UserSyncService.java:
```diff
@@ -274,6 +290,30 @@ private SyncStatusDetail internalSynchronizeStack(Stack stack, UmsUsersState ums
+        LOGGER.info("Syncing Azure Object IDs for environment {}", envCrn);
+
+        UsersState userState = umsUsersState.getUsersState();
+        Map<String, String> usersToAzureObjectIdMap = userState.getUsers().stream()
```
These maps should just be stored in the UmsUsersState directly.

handavid commented on freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/model/FmsGroup.java:
```diff
@@ -1,20 +1,32 @@
 
     private String name;
 
+    private Optional<String> azureObjectId = Optional.empty();
```
see comment on UmsUsersStateProvider for why we can't store this info in this model

handavid commented on freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/model/FmsUser.java:
```diff
@@ -10,6 +11,8 @@
 
     private String lastName;
 
+    private Optional<String> azureObjectId = Optional.empty();
```
see comment on UmsUsersStateProvider for why we can't store this info in this model

keyki commented on datalake/src/main/java/com/sequenceiq/datalake/cm/ClouderaManagerRangerUtil.java:
```diff
@@ -0,0 +1,124 @@
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ClouderaManagerRangerUtil.class);
+
+    private static final String RANGER_SERVICE_NAME = "ranger";
```
Maybe you could include the cluster-cm module and move all the CM related code there and use it from Data Lake.

aarman-cloudera replied:
Do you mind if we look at this in a follow up task? Moving the logic into cluster-cm is straight forward, but pulling in cluster-cm as dependency seems problematic. I ran into dependency conflicts (hibernate, CachingConfig) as well as it requiring additional dependencies (I think it required bringing an implementation of CloudbreakEventService). The  refactoring required might be beyond the scope of this task. I can file a ticket to see if the CM portion can be easily be moved into cluster-cm and client-cm.

keyki commented on datalake/src/main/resources/application.yml:
```diff
@@ -100,4 +100,7 @@ cb:
\ No newline at end of file
+  urls: http://localhost:3000/notifications
+
+clusterProxy:
```
Note: we need to add cluster proxy URL in CBD and helm-charts as well

aarman-cloudera replied:
Good point, I'm looking into it.

keyki commented on freeipa/src/main/resources/application.yml:
```diff
@@ -64,6 +64,9 @@ freeipa:
   environment:
     url: http://localhost:8088
     contextPath: /environmentservice
+  sdx:
```
This is a new dependency of FreeIPA, thus we need to introduce it in CBD and helm-charts as well.

lacikaaa commented on datalake/src/main/java/com/sequenceiq/datalake/cm/CloudIdentityRangerSyncService.java:
```diff
@@ -0,0 +1,41 @@
+    public void setAzureCloudIdentityMapping(String environmentCrn, Map<String, String> azureUserMapping, Map<String, String> azureGroupMapping) {
+        LOGGER.info("Setting Azure cloud id mappings for environment = {}", environmentCrn);
+        List<SdxCluster> sdxClusters = sdxService.listSdxByEnvCrn(environmentCrn);
+        if (sdxClusters.isEmpty()) {
+            LOGGER.info("Environment has no datalake clusters to sync");
+        }
```
instead of logging when empty we should log sdxclusters we will sync, so the `if` won't be necessary

lacikaaa commented on datalake/src/main/java/com/sequenceiq/datalake/cm/CloudIdentityRangerSyncService.java:
```diff
@@ -0,0 +1,41 @@
+            try {
+                clouderaManagerRangerUtil.setAzureCloudIdentityMapping(stackCrn, azureUserMapping, azureGroupMapping);
+            } catch (ApiException e) {
+                throw new RuntimeException("Encountered api exception", e);
```
I think you should use a more specific exception here, and might log it

aarman-cloudera replied:
I've added a new exception type.

lacikaaa commented on datalake/src/main/java/com/sequenceiq/datalake/cm/CloudIdentityRangerSyncService.java:
```diff
@@ -0,0 +1,41 @@
+            LOGGER.info("Environment has no datalake clusters to sync");
+        }
+        sdxClusters.forEach(sdxCluster -> {
+            String stackCrn = sdxCluster.getStackCrn();
+            LOGGER.info("Updating azure cloud id mappings for datalake stack crn = {}, environment = {}", stackCrn, environmentCrn);
+            try {
+                clouderaManagerRangerUtil.setAzureCloudIdentityMapping(stackCrn, azureUserMapping, azureGroupMapping);
+            } catch (ApiException e) {
+                throw new RuntimeException("Encountered api exception", e);
+            }
+        });
```
could be refaactored into a separate method

lacikaaa commented on datalake/src/main/java/com/sequenceiq/datalake/cm/ClouderaManagerProxiedClientFactory.java:
```diff
@@ -0,0 +1,32 @@
+
+    private static final int CM_READ_TIMEOUT_MS =  60 * 1000;
+
+    private static final int CM_CONNECT_TIMEOUT_MS = 20 * 1000;
```
why 20sec? I think we use 1 min in CB for this

keyki replied:
The cm module have these already configured to use proxy and direct connections, I think that should be used instead of creating new CM related code in other modules.

aarman-cloudera replied:
I think we used 20 seconds in thunderhead repo. I've updated this to 60 sec. As mentioned in previous comment, I think it makes sense to move the logic into other modules as a separate ticket.

lacikaaa commented on datalake/src/main/java/com/sequenceiq/datalake/cm/ClouderaManagerProxiedClientFactory.java:
```diff
@@ -0,0 +1,32 @@
+@Component
+public class ClouderaManagerProxiedClientFactory {
+
+    private static final String API_VERSION_32 = "v32";
```
why this version? in CB we use 31 or 40. I think this should be in sync

aarman-cloudera replied:
Updated to 31

lacikaaa commented on datalake/src/main/java/com/sequenceiq/datalake/cm/ClouderaManagerProxiedClientFactory.java:
```diff
@@ -0,0 +1,32 @@
+import java.util.concurrent.TimeUnit;
+
+@Component
+public class ClouderaManagerProxiedClientFactory {
```
altogether it would be nice if you could use https://github.com/hortonworks/cloudbreak/blob/master/client-cm/src/main/java/com/sequenceiq/cloudbreak/cm/client/ClouderaManagerApiClientProvider.java instead of this class

aarman-cloudera replied:
As mentioned in previous comment, can we look into this as a separate ticket.

lacikaaa commented on datalake/src/main/java/com/sequenceiq/datalake/cm/ClouderaManagerRangerUtil.java:
```diff
@@ -0,0 +1,124 @@
+        String clusterName = getClusterName(client);
+        String rangerUserSyncRoleName = getRangerUserSyncRoleName(client, clusterName);
+        RolesResourceApi rolesResourceApi = clouderaManagerApiFactory.getRolesResourceApi(client);
+        if (!supportsCloudIdMapping(rolesResourceApi, clusterName, rangerUserSyncRoleName)) {
+            LOGGER.info("This version of CM does not support cloud identity mapping. Skipping.");
+            return;
+        }
```
please don't use in middle of method return

lacikaaa commented on datalake/src/main/java/com/sequenceiq/datalake/controller/sdx/SdxController.java:
```diff
@@ -239,4 +248,19 @@ public FlowIdentifier stopByCrn(@ResourceCrn String crn) {
+    @Override
+    @DisableCheckPermissions
+    public void setRangerCloudIdentityMapping(String envCrn, SetRangerCloudIdentityMappingRequest request) {
+        checkIsInternalActor();
```
isn't already there a general solution for this @horadla23 ?

horadla23 replied:
yes, there is, last week we introduced `@InternalOnly` annotation, that should be used here instead of `@DisabledCheckPermissions` (I need to update authz readme)

horadla23 replied:
and `@InternalOnly` will guarantee that the method can be called only by internal actor

aarman-cloudera replied:
Thanks, I've switched the new annotation.

lacikaaa commented on datalake/src/main/java/com/sequenceiq/datalake/controller/sdx/SdxController.java:
```diff
@@ -239,4 +248,19 @@ public FlowIdentifier stopByCrn(@ResourceCrn String crn) {
         return cdpConfigService.getAdvertisedRuntimes();
     }
 
+    // custom authorization check
```
unnecessary comment

lacikaaa commented on datalake/src/main/java/com/sequenceiq/datalake/service/sdx/SdxService.java:
```diff
@@ -509,6 +509,12 @@ private void setTagsSafe(SdxClusterRequest sdxClusterRequest, SdxCluster sdxClus
     }
 
+    public List<SdxCluster> listSdxByEnvCrn(String envCrn) {
+        LOGGER.info("Listing SDX clusters by environment crn {}", envCrn);
```
this might be debug if we really need it

lacikaaa commented on datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SetRangerCloudIdentityMappingRequest.java:
```diff
@@ -0,0 +1,27 @@
+
+import java.util.Map;
+
+public class SetRangerCloudIdentityMappingRequest {
```
I think some annotations are missing here, like ignore unknown etc

aarman-cloudera replied:
For now both of these fields are required. I think once we add support for other cloud providers we might want to loosen that requirement.

aarman-cloudera replied:
Thanks! (I don't think I saw similar annotations in the other sdx requets)

lacikaaa commented on datalake/src/main/java/com/sequenceiq/datalake/cm/ClouderaManagerRangerUtil.java:
```diff
@@ -0,0 +1,124 @@
+    }
+
+    @Retryable(value = RetryException.class, maxAttempts = 5, backoff = @Backoff(delay = 500))
+    public void triggerRoleRefresh(ApiClient client, String clusterName, String serviceName, String roleName) {
```
this could be private

aarman-cloudera replied:
Sorry thought I made it private, fixed now.

lacikaaa commented on freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/UmsUsersStateProvider.java:
```diff
@@ -217,11 +219,26 @@ private void addMemberToInternalTrackingGroup(UsersState.Builder usersStateBuild
+        if (azureCloudIdentities.isEmpty()) {
+            return Optional.empty();
+        }
+        if (azureCloudIdentities.size() > 1) {
+            throw new IllegalStateException(String.format("List contains multiple azure cloud identities = %s", cloudIdentities));
+        }
+        String azureObjectId = Iterables.getOnlyElement(azureCloudIdentities).getCloudIdentityName().getAzureCloudIdentityName().getObjectId();
+        return Optional.of(azureObjectId);
```
this should go into an else branch, maybe into a separate method

handavid commented on datalake-api/src/main/java/com/sequenceiq/sdx/api/endpoint/SdxEndpoint.java:
```diff
@@ -165,4 +166,10 @@ SdxClusterResponse create(@ValidStackNameFormat @ValidStackNameLength @PathParam
     List<AdvertisedRuntime> advertisedRuntimes();
 
+    @POST
+    @Path("/crn/{envCrn}/ranger_cloud_identity_mapping")
```
See the path on line 68. it looks like
/envcrn/{envCrn}
This should match to distinguish between the resource crn and the environment crn

handavid commented on freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/UserSyncService.java:
```diff
@@ -274,6 +292,51 @@ private SyncStatusDetail internalSynchronizeStack(Stack stack, UmsUsersState ums
+        String envCrn = stack.getEnvironmentCrn();
+        LOGGER.info("Syncing Azure Object IDs for environment {}", envCrn);
+
+        Map<String, String> userToAzureObjectIdMap = getAzureObjectIdMap(umsUsersState.getUserToCloudIdentityMap());
```
this adds cloud identities for all users, regardless of if they have access to the environment. Is this desirable? I would think that you would only care about users that have access to the environment. Imagine if an account has 1000 user but the environment only has 100. You would be syncing the other 900 users cloud identities.

You can filter this map by the users that are in the umsUsersState.getUserState().getUsers()

aarman-cloudera replied:
Good point, I've added the filtering part. Also I've moved most of this logic out of UserSyncService and into CloudIdentitySyncService.

handavid commented on freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/UserSyncService.java:
```diff
@@ -283,6 +290,12 @@ private SyncStatusDetail internalSynchronizeStack(Stack stack, UmsUsersState ums
 
+            boolean hasCloudIdentityEntitlement = entitlementService.cloudIdentityMappingEnabled(INTERNAL_ACTOR_CRN, stack.getAccountId());
+            // TODO For now we only sync cloud ids during full sync. We should eventually allow more granular syncs (actor level and group level sync).
+            if (hasCloudIdentityEntitlement && fullSync) {
```
nit: if you reverse the order of conditions and inline the entitlement check here, then we can skip the entitlement check if it's not a full sync.

aarman-cloudera replied:
nice one

lacikaaa commented on freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/CloudIdentitySyncService.java:
```diff
@@ -0,0 +1,99 @@
+        }
+    }
+
+    private Map<String, String> getAzureObjectIdMap(Map<String, List<CloudIdentity>> cloudIdentityMapping) {
```
some log here and in `getOptionalAzureObjectId` might be useful to investigate if something is not as desired

lacikaaa commented on freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/CloudIdentitySyncService.java:
```diff
@@ -0,0 +1,99 @@
+        if (azureCloudIdentities.isEmpty()) {
+            return Optional.empty();
+        } else if (azureCloudIdentities.size() > 1) {
+            throw new IllegalStateException(String.format("List contains multiple azure cloud identities = %s", cloudIdentities));
```
shouldn't we handle this exception somewhere? or is it ok if `UserSyncService#internalSynchronizeStack` handles it? 
cc @handavid

handavid replied:
good point. I think we can expand the try block up on line 50 to start on line 42 instead. That will catch this exception and register a warning.

aarman-cloudera replied:
I don't think we need to fail super cleanly in this situation. This is a very unexpected scenario.

aarman-cloudera replied:
That said, I've taken David's advice.

sodre90 commented on datalake/src/main/java/com/sequenceiq/datalake/cm/ClouderaManagerRangerUtil.java:
```diff
@@ -0,0 +1,123 @@
+        roleNameList.addItemsItem(roleName);
+        RoleCommandsResourceApi roleCommandsResourceApi = clouderaManagerApiFactory.getRoleCommandsResourceApi(client);
+        try {
+            ApiBulkCommandList bulkResponse = roleCommandsResourceApi.refreshCommand(clusterName, serviceName, roleNameList);
```
Good point, it might be better to do role refresh in a flow.

aarman-cloudera replied:
This just triggers role refresh, it doesn't wait around for the role refresh to succeed, so FreeIPA won't hang. I somewhat described this in the commit description. So in a follow up commit we'll want to take the ApiCommand response and poll the status of the command. So the async part will come in a follow up commit (I've discussed with Mark and David and we weren't considering using flow, but maybe we should).

handavid replied:
The actual activity is tracked in CM. I don't think we even need a poller in the datalake service and can just let the FMS poll the datalake service. We can add a poller if we have a use-case where the datalake service needs to push the completion status to another service.

keyki commented:
Should we also add the entitlement to the mock UMS service so we can test this locally?

aarman-cloudera commented:
> Should we also add the entitlement to the mock UMS service so we can test this locally?

That makes sense, I've updated mock UMS

keyki commented on mock-caas/src/main/java/com/sequenceiq/caas/grpc/service/auth/MockUserManagementService.java:
```diff
@@ -531,6 +533,7 @@ public void getAccount(GetAccountRequest request, StreamObserver<GetAccountRespo
                                 .setWorkloadSubdomain(ACCOUNT_SUBDOMAIN)
                                 .addEntitlements(createEntitlement(CDP_AZURE))
                                 .addEntitlements(createEntitlement(CDP_AUTOMATIC_USERSYNC_POLLER))
+                                .addEntitlements(createEntitlement(CDP_CLOUD_IDENTITY_MAPPING))
```
I think we should have an env var to enable and disable the entitlement. See a couple of lines above.

aarman-cloudera commented on datalake/src/main/java/com/sequenceiq/datalake/controller/exception/RangerCloudIdentitySyncException.java:
```diff
@@ -0,0 +1,12 @@
+package com.sequenceiq.datalake.controller.exception;
+
+public class RangerCloudIdentitySyncException extends RuntimeException {
```
So the semantics of the exception might change once we add polling of the refresh command in follow on commit. So I think we can hold off on an exception mapper until then.

aarman-cloudera commented:
> some of my comments are not resolved/answered. until they are I would like to keep the review state in changes requested

Sorry just noticed them (they got buried under hidden). I've addressed those comments, please check again.
